File: scripts/check-south-africa.py
"""Validate isolated South Africa public files without network or external packages."""
from pathlib import Path
from html.parser import HTMLParser
from urllib.parse import urlsplit, unquote
import json
import re
import xml.etree.ElementTree as ET

ROOT = Path(__file__).resolve().parents[1] / 'south-africa'
BASE = 'https://sa.gloriatrading.com/'
FORBIDDEN = re.compile(
    r'preferential Import Permit|will be accepted by Uber|industry-leading|number one exporter|best quality|unmatched experience',
    re.I,
)

class Page(HTMLParser):
    def __init__(self):
        super().__init__()
        self.refs = []
        self.canonical = []
        self.ids = []
        self.h1 = 0
        self.text = []
    def handle_starttag(self, tag, attrs):
        attrs = dict(attrs)
        if tag == 'h1':
            self.h1 += 1
        if 'id' in attrs:
            self.ids.append(attrs['id'])
        for key in ('src', 'href'):
            if attrs.get(key):
                self.refs.append(attrs[key])
        if tag == 'link' and attrs.get('rel') == 'canonical':
            self.canonical.append(attrs['href'])
    def handle_data(self, data):
        self.text.append(data)

html_pages = list(ROOT.glob('*.html'))
assert html_pages, 'No South Africa HTML pages'
for path in html_pages:
    raw = path.read_text(encoding='utf-8')
    parser = Page()
    parser.feed(raw)
    assert parser.h1 == 1, path
    assert len(parser.ids) == len(set(parser.ids)), path
    assert len(parser.canonical) == 1 and parser.canonical[0].startswith(BASE), path
    assert 'sa.gloriatrading.com' in raw
    assert FORBIDDEN.search(raw) is None, (path, FORBIDDEN.search(raw))
    assert 'Import requirements must be confirmed before shipment' in raw, path
    for ref in parser.refs:
        url = urlsplit(ref)
        if url.scheme or url.netloc:
            continue
        target = (path.parent / unquote(url.path)).resolve() if url.path else path
        assert target.is_relative_to(ROOT), (path, ref)
        assert target.exists(), (path, ref)
        if url.fragment:
            p = Page()
            p.feed(target.read_text(encoding='utf-8'))
            assert unquote(url.fragment) in p.ids, (path, ref)

vehicles = json.loads((ROOT / 'data/vehicles.json').read_text(encoding='utf-8'))['vehicles']
ids = set()
for v in vehicles:
    assert v.get('ref_id') and v['ref_id'] not in ids
    ids.add(v['ref_id'])
    assert v.get('make') and v.get('model')
    assert v.get('listing_type') in ('sample', 'available')
    assert v.get('status') in ('published', 'draft')
    for key in ('year', 'mileage_km', 'battery', 'range', 'reference_price_usd', 'video_url'):
        value = v.get(key)
        assert value in (None, '', [])
    for src in v.get('gallery', []):
        assert src.startswith('images/') and '..' not in src
        assert (ROOT / src).is_file(), src

js = (ROOT / 'js/sa.js').read_text(encoding='utf-8')
assert 'wa.me/819076671825' in js
assert 'info@gloriatrading.com' in js
assert 'textContent' in js

for loc in ET.parse(ROOT / 'sitemap.xml').iter('{http://www.sitemaps.org/schemas/sitemap/0.9}loc'):
    assert loc.text.startswith(BASE)
    assert (ROOT / (urlsplit(loc.text).path.lstrip('/') or 'index.html')).is_file()

robots = (ROOT / 'robots.txt').read_text(encoding='utf-8')
assert 'https://sa.gloriatrading.com/sitemap.xml' in robots

print(f'SA checks passed: {len(html_pages)} pages, {len(vehicles)} example types.')

File: scripts/check-south-africa.js
'use strict';
const fs = require('fs');
const path = require('path');

const ROOT = path.resolve(__dirname, '..', 'south-africa');
const BASE = 'https://sa.gloriatrading.com/';
const FORBIDDEN = /preferential Import Permit|will be accepted by Uber|industry-leading|number one exporter|best quality|unmatched experience/i;

function walkHtml(dir) {
  return fs.readdirSync(dir).filter(name => name.endsWith('.html')).map(name => path.join(dir, name));
}

function hrefs(html) {
  const refs = [];
  const re = /\s(?:src|href)="([^"]+)"/g;
  let match;
  while ((match = re.exec(html))) refs.push(match[1]);
  return refs;
}

const pages = walkHtml(ROOT);
if (!pages.length) throw new Error('No South Africa HTML pages');
for (const file of pages) {
  const raw = fs.readFileSync(file, 'utf8');
  if ((raw.match(/<h1[\s>]/g) || []).length !== 1) throw new Error(`h1 count ${file}`);
  if (!raw.includes('rel="canonical"') || !raw.includes(BASE)) throw new Error(`canonical ${file}`);
  if (!raw.includes('sa.gloriatrading.com')) throw new Error(`canonical host ${file}`);
  if (FORBIDDEN.test(raw)) throw new Error(`forbidden claim ${file}`);
  if (!raw.includes('Import requirements must be confirmed before shipment')) {
    throw new Error(`import wording ${file}`);
  }
  const ids = [...raw.matchAll(/\sid="([^"]+)"/g)].map(m => m[1]);
  if (new Set(ids).size !== ids.length) throw new Error(`duplicate id ${file}`);
  for (const ref of hrefs(raw)) {
    if (/^[a-z]+:/i.test(ref) || ref.startsWith('//') || ref.startsWith('#')) continue;
    const target = path.resolve(path.dirname(file), decodeURIComponent(ref.split('#')[0] || '.'));
    const rel = path.relative(ROOT, target);
    if (rel.startsWith('..')) throw new Error(`escape ${file} ${ref}`);
    if (!fs.existsSync(target)) throw new Error(`missing ${file} ${ref}`);
  }
}

const vehicles = JSON.parse(fs.readFileSync(path.join(ROOT, 'data', 'vehicles.json'), 'utf8')).vehicles;
const ids = new Set();
for (const v of vehicles) {
  if (!v.ref_id || ids.has(v.ref_id) || !v.make || !v.model) throw new Error(`vehicle ${v.ref_id}`);
  ids.add(v.ref_id);
  if (v.listing_type !== 'sample' && v.listing_type !== 'available') throw new Error(`listing_type ${v.ref_id}`);
  if (v.status !== 'published' && v.status !== 'draft') throw new Error(`status ${v.ref_id}`);
  for (const key of ['year', 'mileage_km', 'battery', 'range', 'reference_price_usd', 'video_url']) {
    const value = v[key];
    if (!(value === null || value === '' || (Array.isArray(value) && value.length === 0))) {
      throw new Error(`invented field ${v.ref_id} ${key}`);
    }
  }
  for (const src of v.gallery || []) {
    if (!src.startsWith('images/') || src.includes('..')) throw new Error(src);
    if (!fs.existsSync(path.join(ROOT, src))) throw new Error(src);
  }
}

const js = fs.readFileSync(path.join(ROOT, 'js', 'sa.js'), 'utf8');
if (!js.includes('Example vehicle — not current stock')) throw new Error('sample wording');
if (!fs.existsSync(path.join(ROOT, 'admin', 'index.php'))) throw new Error('admin missing');
if (!fs.existsSync(path.join(ROOT, 'lib', 'admin-auth.php'))) throw new Error('admin auth missing');
if (fs.existsSync(path.join(ROOT, 'admin-password.hash')) || fs.existsSync(path.join(ROOT, 'lib', 'admin-password.hash'))) {
  throw new Error('password hash must not be in the public tree');
}

const sitemap = fs.readFileSync(path.join(ROOT, 'sitemap.xml'), 'utf8');
const locs = [...sitemap.matchAll(/<loc>([^<]+)<\/loc>/g)].map(m => m[1]);
if (!locs.length) throw new Error('empty sitemap');
for (const loc of locs) {
  if (!loc.startsWith(BASE)) throw new Error(loc);
  const file = loc.slice(BASE.length) || 'index.html';
  if (!fs.existsSync(path.join(ROOT, file))) throw new Error(loc);
}

const robots = fs.readFileSync(path.join(ROOT, 'robots.txt'), 'utf8');
if (!robots.includes('https://sa.gloriatrading.com/sitemap.xml')) throw new Error('robots');

console.log(`SA checks passed: ${pages.length} pages, ${vehicles.length} example types.`);
